File: androidApp/src/main/kotlin/org/coreypett/fullstack/features/marketdetails/components/MarketChartSection.kt
package org.coreypett.fullstack.features.marketdetails.components

import android.view.ViewGroup
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.graphics.toArgb
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.ui.viewinterop.AndroidView
import com.tradingview.lightweightcharts.api.chart.models.color.IntColor
import com.tradingview.lightweightcharts.api.chart.models.color.surface.SolidColor
import com.tradingview.lightweightcharts.api.interfaces.SeriesApi
import com.tradingview.lightweightcharts.api.options.models.CandlestickSeriesOptions
import com.tradingview.lightweightcharts.api.options.models.ChartOptions
import com.tradingview.lightweightcharts.api.options.models.GridLineOptions
import com.tradingview.lightweightcharts.api.options.models.GridOptions
import com.tradingview.lightweightcharts.api.options.models.HistogramSeriesOptions
import com.tradingview.lightweightcharts.api.options.models.LayoutOptions
import com.tradingview.lightweightcharts.api.options.models.LocalizationOptions
import com.tradingview.lightweightcharts.api.options.models.PriceScaleMargins
import com.tradingview.lightweightcharts.api.options.models.PriceScaleOptions
import com.tradingview.lightweightcharts.api.options.models.TimeScaleOptions
import com.tradingview.lightweightcharts.api.series.common.SeriesData
import com.tradingview.lightweightcharts.api.series.models.CandlestickData
import com.tradingview.lightweightcharts.api.series.models.HistogramData
import com.tradingview.lightweightcharts.api.series.models.PriceFormat
import com.tradingview.lightweightcharts.api.series.models.PriceScaleId
import com.tradingview.lightweightcharts.api.series.models.Time
import com.tradingview.lightweightcharts.runtime.plugins.Eval
import com.tradingview.lightweightcharts.view.ChartsView
import org.coreypett.fullstack.features.marketdetails.MarketDetailsChartIntervals
import org.coreypett.fullstack.market.MR
import org.coreypett.fullstack.market.candle.model.CandleBar
import org.coreypett.fullstack.market.candle.model.CandleInterval
import org.coreypett.fullstack.market.candle.presentation.CandleChartStatus
import org.coreypett.fullstack.market.candle.presentation.CandleChartState
import org.coreypett.fullstack.support.toComposeColor

@Composable
fun MarketChartSection(
    selectedInterval: CandleInterval,
    candleState: CandleChartState,
    onIntervalSelected: (CandleInterval) -> Unit,
) {
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .background(MR.colors.app_background.toComposeColor()),
        verticalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        IntervalSelector(
            selectedInterval = selectedInterval,
            onIntervalSelected = onIntervalSelected,
        )
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .height(260.dp),
        ) {
            Image(
                painter = painterResource(id = MR.images.fullstack_logo.drawableResId),
                contentDescription = null,
                modifier = Modifier
                    .align(Alignment.Center)
                    .size(150.dp)
                    .alpha(0.18f),
            )
            if (candleState.bars.isEmpty()) {
                if (candleState.status == CandleChartStatus.Failed) {
                    FillEmptyState(
                        message = candleState.message ?: "Loading chart",
                        modifier = Modifier.fillMaxSize(),
                    )
                } else {
                    ChartSkeleton(modifier = Modifier.fillMaxSize())
                }
            } else {
                LightweightCandleChart(
                    bars = candleState.bars,
                    modifier = Modifier.fillMaxSize(),
                )
            }
        }
    }
}

@Composable
private fun ChartSkeleton(
    modifier: Modifier = Modifier,
) {
    BoxWithConstraints(modifier = modifier) {
        val spacing = 8.dp
        val chartHeight = maxHeight
        val barWidth = ((maxWidth - 32.dp - (spacing * (ChartSkeletonFractions.size - 1))) / ChartSkeletonFractions.size)
            .coerceAtLeast(8.dp)

        Row(
            modifier = Modifier
                .fillMaxSize()
                .padding(horizontal = 16.dp, vertical = 18.dp),
            verticalAlignment = Alignment.Bottom,
            horizontalArrangement = Arrangement.spacedBy(spacing),
        ) {
            ChartSkeletonFractions.forEachIndexed { index, fraction ->
                Column(
                    modifier = Modifier
                        .weight(1f)
                        .fillMaxHeight(),
                    horizontalAlignment = Alignment.CenterHorizontally,
                    verticalArrangement = Arrangement.Bottom,
                ) {
                    SkeletonBlock(
                        width = 2.dp,
                        height = chartHeight * 0.22f,
                        cornerRadius = 1.dp,
                    )
                    Spacer(Modifier.height(4.dp))
                    SkeletonBlock(
                        width = barWidth,
                        height = (chartHeight * fraction * 0.42f).coerceAtLeast(28.dp),
                        cornerRadius = 3.dp,
                        modifier = Modifier.alpha(if (index % 2 == 0) 0.86f else 0.64f),
                    )
                }
            }
        }
    }
}

@Composable
private fun IntervalSelector(
    selectedInterval: CandleInterval,
    onIntervalSelected: (CandleInterval) -> Unit,
) {
    Row(
        modifier = Modifier.fillMaxWidth(),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(12.dp),
    ) {
        Text(
            text = "Int.",
            color = MR.colors.text_primary.toComposeColor(),
            fontSize = 12.sp,
            fontWeight = FontWeight.Medium,
            maxLines = 1,
        )
        Row(
            horizontalArrangement = Arrangement.spacedBy(12.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            MarketDetailsChartIntervals.forEach { interval ->
                val isSelected = interval == selectedInterval
                Text(
                    text = interval.chartLabel,
                    color = if (isSelected) {
                        MR.colors.text_primary.toComposeColor()
                    } else {
                        MR.colors.control_text_dimmed.toComposeColor()
                    },
                    fontSize = 12.sp,
                    fontWeight = if (isSelected) FontWeight.SemiBold else FontWeight.Medium,
                    maxLines = 1,
                    modifier = Modifier.clickable { onIntervalSelected(interval) },
                )
            }
        }
    }
}

@Composable
private fun LightweightCandleChart(
    bars: List<CandleBar>,
    modifier: Modifier = Modifier,
) {
    val backgroundColor = android.graphics.Color.TRANSPARENT
    val textColor = MR.colors.text_tertiary.toComposeColor().toArgb()
    val gridColor = MR.colors.app_border.toComposeColor().copy(alpha = 0.36f).toArgb()
    val bidColor = MR.colors.bid.toComposeColor().toArgb()
    val askColor = MR.colors.ask.toComposeColor().toArgb()
    val volumeBidColor = MR.colors.bid.toComposeColor().copy(alpha = 0.26f).toArgb()
    val volumeAskColor = MR.colors.ask.toComposeColor().copy(alpha = 0.26f).toArgb()
    val holder = remember { CandleChartHolder() }
    val priceFormatter = remember { Eval(DollarPriceFormatterJavaScript) }
    val priceFormat = remember(priceFormatter) {
        PriceFormat.priceFormatCustom(formatter = priceFormatter, minMove = 0.01f)
    }
    val candleData = remember(bars, bidColor, askColor) {
        bars.map { bar ->
            bar.toCandlestickData(
                upColor = bidColor,
                downColor = askColor,
            )
        }
    }
    val volumeData = remember(bars, volumeBidColor, volumeAskColor) {
        bars.map { bar ->
            bar.toVolumeData(
                upColor = volumeBidColor,
                downColor = volumeAskColor,
            )
        }
    }

    AndroidView(
        modifier = modifier,
        factory = { context ->
            ChartsView(context).apply {
                layoutParams = ViewGroup.LayoutParams(
                    ViewGroup.LayoutParams.MATCH_PARENT,
                    ViewGroup.LayoutParams.MATCH_PARENT,
                )
                setBackgroundColor(backgroundColor)
                subscribeOnChartStateChange { chartState ->
                    if (chartState is ChartsView.State.Ready && holder.candleSeries == null) {
                        api.applyOptions(
                            ChartOptions(
                                layout = LayoutOptions(
                                    background = SolidColor(backgroundColor),
                                    textColor = IntColor(textColor),
                                    fontSize = 11,
                                ),
                                localization = LocalizationOptions(
                                    priceFormatter = priceFormatter,
                                ),
                                rightPriceScale = PriceScaleOptions(
                                    autoScale = true,
                                    scaleMargins = PriceScaleMargins(top = 0.08f, bottom = 0.28f),
                                    borderVisible = false,
                                    alignLabels = true,
                                ),
                                grid = GridOptions(
                                    vertLines = GridLineOptions(
                                        color = IntColor(gridColor),
                                        visible = true,
                                    ),
                                    horzLines = GridLineOptions(
                                        color = IntColor(gridColor),
                                        visible = true,
                                    ),
                                ),
                                timeScale = TimeScaleOptions(
                                    borderVisible = false,
                                    timeVisible = true,
                                    secondsVisible = false,
                                    rightOffset = 4f,
                                    barSpacing = 8f,
                                    minBarSpacing = 5f,
                                    shiftVisibleRangeOnNewBar = true,
                                ),
                            ),
                        )
                        api.addCandlestickSeries(
                            CandlestickSeriesOptions(
                                upColor = IntColor(bidColor),
                                downColor = IntColor(askColor),
                                borderVisible = false,
                                borderUpColor = IntColor(bidColor),
                                borderDownColor = IntColor(askColor),
                                wickUpColor = IntColor(bidColor),
                                wickDownColor = IntColor(askColor),
                                priceFormat = priceFormat,
                            ),
                        ) { series ->
                            holder.candleSeries = series
                            series.setData(holder.pendingCandleData.ifEmpty { candleData })
                            api.timeScale.scrollToRealTime()
                        }
                        api.addHistogramSeries(
                            HistogramSeriesOptions(
                                lastValueVisible = false,
                                priceLineVisible = false,
                                priceScaleId = PriceScaleId(VolumePriceScaleId),
                                priceFormat = PriceFormat.priceFormatBuiltIn(
                                    type = PriceFormat.Type.VOLUME,
                                    precision = 0,
                                    minMove = 1f,
                                ),
                            ),
                        ) { volumeSeries ->
                            holder.volumeSeries = volumeSeries
                            volumeSeries.priceScale().applyOptions(
                                PriceScaleOptions(
                                    scaleMargins = PriceScaleMargins(top = 0.78f, bottom = 0f),
                                    borderVisible = false,
                                    visible = false,
                                ),
                            )
                            volumeSeries.setData(holder.pendingVolumeData.ifEmpty { volumeData })
                        }
                    }
                }
            }
        },
        update = { chartView ->
            holder.pendingCandleData = candleData
            holder.pendingVolumeData = volumeData
            val candleSeries = holder.candleSeries
            if (candleSeries != null && candleData.isNotEmpty()) {
                candleSeries.setData(candleData)
                holder.volumeSeries?.setData(volumeData)
                chartView.api.timeScale.scrollToRealTime()
            }
        },
    )
}

private class CandleChartHolder {
    var candleSeries: SeriesApi? = null
    var volumeSeries: SeriesApi? = null
    var pendingCandleData: List<SeriesData> = emptyList()
    var pendingVolumeData: List<SeriesData> = emptyList()
}

private fun CandleBar.toCandlestickData(
    upColor: Int,
    downColor: Int,
): SeriesData {
    val candleColor = if (close >= open) upColor else downColor
    return CandlestickData(
        time = Time.Utc(openTimeMillis / 1_000L),
        open = open.toFloat(),
        high = high.toFloat(),
        low = low.toFloat(),
        close = close.toFloat(),
        color = IntColor(candleColor),
        borderColor = IntColor(candleColor),
        wickColor = IntColor(candleColor),
    )
}

private fun CandleBar.toVolumeData(
    upColor: Int,
    downColor: Int,
): SeriesData = HistogramData(
    time = Time.Utc(openTimeMillis / 1_000L),
    value = volume.toFloat(),
    color = IntColor(if (close >= open) upColor else downColor),
)

private val CandleInterval.chartLabel: String
    get() = if (this == CandleInterval.OneDay) "1D" else displayName

private const val VolumePriceScaleId = "volume"

private val ChartSkeletonFractions = listOf(0.42f, 0.66f, 0.54f, 0.76f, 0.48f, 0.58f, 0.82f, 0.62f, 0.44f, 0.70f, 0.52f, 0.64f)

private val DollarPriceFormatterJavaScript = """
function(price) {
    var sign = price < 0 ? '-$' : '$';
    var rounded = Math.round(Math.abs(price) * 100);
    var whole = String(Math.floor(rounded / 100)).replace(/\B(?=(\d{3})+(?!\d))/g, ',');
    var cents = rounded % 100;
    if (cents === 0) {
        return sign + whole;
    }
    return sign + whole + '.' + (cents < 10 ? '0' + cents : String(cents));
}
""".trimIndent()

File: iosApp/iosApp/Features/MarketDetails/Components/MarketChartSection.swift
@preconcurrency import LightweightCharts
import Pow
@preconcurrency import SharedLogic
import SwiftUI
import UIKit

struct MarketChartBar {
    let time: Time
    let open: Double
    let high: Double
    let low: Double
    let close: Double
    let volume: Double

    init(shared: CandleBar) {
        self.time = .utc(timestamp: Double(shared.openTimeMillis) / 1_000)
        self.open = shared.open
        self.high = shared.high
        self.low = shared.low
        self.close = shared.close
        self.volume = shared.volume
    }

    init(
        time: Time,
        open: Double,
        high: Double,
        low: Double,
        close: Double,
        volume: Double
    ) {
        self.time = time
        self.open = open
        self.high = high
        self.low = low
        self.close = close
        self.volume = volume
    }

    static let preview: [MarketChartBar] = [
        MarketChartBar(time: .string("2026-06-01"), open: 68120, high: 68980, low: 67840, close: 68720, volume: 48920),
        MarketChartBar(time: .string("2026-06-02"), open: 68720, high: 69410, low: 68240, close: 69080, volume: 53110),
        MarketChartBar(time: .string("2026-06-03"), open: 69080, high: 69860, low: 68620, close: 69650, volume: 61640),
        MarketChartBar(time: .string("2026-06-04"), open: 69650, high: 70180, low: 69140, close: 69320, volume: 45875),
        MarketChartBar(time: .string("2026-06-05"), open: 69320, high: 69940, low: 68780, close: 69790, volume: 50280),
        MarketChartBar(time: .string("2026-06-06"), open: 69790, high: 70420, low: 69440, close: 70130, volume: 72860),
        MarketChartBar(time: .string("2026-06-07"), open: 70130, high: 70680, low: 69690, close: 69910, volume: 69410),
        MarketChartBar(time: .string("2026-06-08"), open: 69910, high: 70220, low: 69120, close: 69480, volume: 56300),
        MarketChartBar(time: .string("2026-06-09"), open: 69480, high: 70010, low: 68980, close: 69870, volume: 47725),
        MarketChartBar(time: .string("2026-06-10"), open: 69870, high: 70720, low: 69610, close: 70420, volume: 81240),
        MarketChartBar(time: .string("2026-06-11"), open: 70420, high: 70980, low: 69920, close: 70240, volume: 63880),
        MarketChartBar(time: .string("2026-06-12"), open: 70240, high: 70620, low: 69380, close: 69690, volume: 58540),
        MarketChartBar(time: .string("2026-06-13"), open: 69690, high: 70130, low: 69160, close: 69980, volume: 42935),
        MarketChartBar(time: .string("2026-06-14"), open: 69980, high: 70340, low: 68720, close: 69125.75, volume: 76520),
    ]
}

struct MarketChartSection: View {
    let market: MarketSymbol
    let selectedInterval: CandleInterval
    let bars: [MarketChartBar]
    let isLoading: Bool
    let onSelectInterval: (CandleInterval) -> Void

    var body: some View {
        VStack(alignment: .leading, spacing: 10) {
            IntervalSelector(
                selectedInterval: selectedInterval,
                onSelect: onSelectInterval
            )
            .padding(.horizontal, 16)

            ZStack {
                ChartWatermark(market: market)

                if isLoading {
                    ChartSkeleton()
                        .transition(.movingParts.blur.combined(with: .opacity))
                } else {
                    MarketCandlestickChart(bars: bars)
                        .transition(.movingParts.blur.combined(with: .opacity))
                }
            }
                .frame(height: 260)
                .animation(.easeInOut(duration: 0.22), value: isLoading)
        }
    }
}

private struct ChartWatermark: View {
    let market: MarketSymbol

    var body: some View {
        VStack {
            if let image = MR.images.shared.fullstack_logo.toUIImage() {
                Image(uiImage: image)
                    .resizable()
                    .scaledToFit()
                    .opacity(0.18)
            }
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .allowsHitTesting(false)
    }
}

private struct IntervalSelector: View {
    let selectedInterval: CandleInterval
    let onSelect: (CandleInterval) -> Void

    var body: some View {
        HStack(spacing: 12) {
            Text("Int.")
                .font(.system(size: 12, weight: .medium))
                .foregroundStyle(MR.colors.shared.text_primary.swiftUIColor)
                .lineLimit(1)

            HStack(spacing: 12) {
                ForEach(chartIntervals, id: \.self) { interval in
                    Button {
                        onSelect(interval)
                    } label: {
                        Text(intervalLabel(for: interval))
                            .font(.system(size: 12, weight: selectedInterval == interval ? .semibold : .medium))
                            .foregroundStyle(intervalColor(for: interval))
                            .lineLimit(1)
                    }
                    .buttonStyle(.plain)
                }
            }
        }
    }

    private func intervalColor(for interval: CandleInterval) -> Color {
        selectedInterval == interval ? MR.colors.shared.control_text_selected.swiftUIColor : MR.colors.shared.control_text_dimmed.swiftUIColor
    }
}

private struct ChartSkeleton: View {
    private let bars: [CGFloat] = [0.42, 0.66, 0.54, 0.76, 0.48, 0.58, 0.82, 0.62, 0.44, 0.7, 0.52, 0.64]

    var body: some View {
        GeometryReader { geometry in
            let width = geometry.size.width
            let height = geometry.size.height
            let spacing: CGFloat = 8
            let barWidth = max((width - CGFloat(bars.count - 1) * spacing - 32) / CGFloat(bars.count), 8)

            HStack(alignment: .bottom, spacing: spacing) {
                ForEach(Array(bars.enumerated()), id: \.offset) { index, fraction in
                    VStack(spacing: 4) {
                        SkeletonBlock(width: 2, height: height * 0.22, cornerRadius: 1)

                        SkeletonBlock(
                            width: barWidth,
                            height: max(height * fraction * 0.42, 28),
                            cornerRadius: 3
                        )
                        .opacity(index.isMultiple(of: 2) ? 0.86 : 0.64)
                    }
                    .frame(maxHeight: .infinity, alignment: .bottom)
                }
            }
            .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .bottom)
            .padding(.horizontal, 16)
            .padding(.vertical, 18)
        }
    }
}

private struct MarketCandlestickChart: UIViewRepresentable {
    let bars: [MarketChartBar]

    func makeCoordinator() -> Coordinator {
        Coordinator()
    }

    func makeUIView(context: Context) -> LightweightCharts {
        let chart = LightweightCharts(options: chartOptions)
        chart.isOpaque = false
        chart.backgroundColor = .clear
        chart.clearWebViewBackground()

        let series = chart.addCandlestickSeries(options: seriesOptions)
        context.coordinator.series = series
        let volumeSeries = chart.addHistogramSeries(options: volumeSeriesOptions)
        volumeSeries.priceScale().applyOptions(options: volumePriceScaleOptions)
        context.coordinator.volumeSeries = volumeSeries
        series.setData(data: bars.map(\.candlestickData))
        volumeSeries.setData(data: bars.map(\.volumeData))
        chart.timeScale().scrollToRealTime()

        return chart
    }

    func updateUIView(_ chart: LightweightCharts, context: Context) {
        chart.isOpaque = false
        chart.backgroundColor = .clear
        chart.clearWebViewBackground()
        context.coordinator.series?.setData(data: bars.map(\.candlestickData))
        context.coordinator.volumeSeries?.setData(data: bars.map(\.volumeData))
        chart.timeScale().scrollToRealTime()
    }

    final class Coordinator {
        var series: CandlestickSeries?
        var volumeSeries: HistogramSeries?
    }

    private var chartOptions: ChartOptions {
        marketChartOptions()
    }

    private var seriesOptions: CandlestickSeriesOptions {
        marketChartSeriesOptions()
    }

    private var volumeSeriesOptions: HistogramSeriesOptions {
        marketChartVolumeSeriesOptions()
    }
}

private extension MarketChartBar {
    var candlestickData: CandlestickData {
        CandlestickData(time: time, open: open, high: high, low: low, close: close)
    }

    var volumeData: HistogramData {
        HistogramData(color: volumeColor, time: time, value: volume)
    }

    private var volumeColor: ChartColor {
        close >= open ? MR.colors.shared.bid.swiftUIColor.opacity(0.26).chartColor : MR.colors.shared.ask.swiftUIColor.opacity(0.26).chartColor
    }
}

private func marketChartOptions() -> ChartOptions {
    ChartOptions(
        autoSize: true,
        layout: LayoutOptions(
            background: .solid(color: Color.clear.chartColor),
            textColor: MR.colors.shared.text_secondary.swiftUIColor.chartColor,
            attributionLogo: false
        ),
        rightPriceScale: VisiblePriceScaleOptions(
            scaleMargins: PriceScaleMargins(top: 0.08, bottom: 0.28),
            borderVisible: false
        ),
        timeScale: TimeScaleOptions(borderVisible: false),
        crosshair: CrosshairOptions(mode: .normal),
        grid: GridOptions(
            verticalLines: GridLineOptions(color: MR.colors.shared.app_border.swiftUIColor.opacity(0.36).chartColor),
            horizontalLines: GridLineOptions(color: MR.colors.shared.app_border.swiftUIColor.opacity(0.36).chartColor)
        )
    )
}

private func marketChartVolumeSeriesOptions() -> HistogramSeriesOptions {
    HistogramSeriesOptions(
        lastValueVisible: false,
        priceScaleId: "volume",
        priceLineVisible: false,
        priceFormat: .builtIn(BuiltInPriceFormat(type: .volume, precision: nil, minMove: nil)),
        color: MR.colors.shared.text_tertiary.swiftUIColor.opacity(0.22).chartColor
    )
}

private var volumePriceScaleOptions: PriceScaleOptions {
    PriceScaleOptions(
        scaleMargins: PriceScaleMargins(top: 0.78, bottom: 0),
        borderVisible: false,
        visible: false
    )
}

private func marketChartSeriesOptions() -> CandlestickSeriesOptions {
    CandlestickSeriesOptions(
        lastValueVisible: true,
        priceLineVisible: true,
        priceFormat: marketChartPriceFormat(),
        upColor: MR.colors.shared.bid.swiftUIColor.chartColor,
        downColor: MR.colors.shared.ask.swiftUIColor.chartColor,
        borderVisible: false,
        wickUpColor: MR.colors.shared.bid.swiftUIColor.chartColor,
        wickDownColor: MR.colors.shared.ask.swiftUIColor.chartColor
    )
}

private func marketChartPriceFormat() -> PriceFormat {
    .custom(
        CustomPriceFormat(
            minMove: 0.01,
            formatterJavaScript: dollarPriceFormatterJavaScript,
            tickmarksFormatterJavaScript: dollarTickmarksFormatterJavaScript
        )
    )
}

private let dollarPriceFormatterJavaScript = #"""
function(price) {
    var sign = price < 0 ? '-$' : '$';
    var rounded = Math.round(Math.abs(price) * 100);
    var whole = String(Math.floor(rounded / 100)).replace(/\B(?=(\d{3})+(?!\d))/g, ',');
    var cents = rounded % 100;
    if (cents === 0) {
        return sign + whole;
    }
    return sign + whole + '.' + (cents < 10 ? '0' + cents : String(cents));
}
"""#

private let dollarTickmarksFormatterJavaScript = #"""
function(prices) {
    return prices.map(function(price) {
        var sign = price < 0 ? '-$' : '$';
        var rounded = Math.round(Math.abs(price) * 100);
        var whole = String(Math.floor(rounded / 100)).replace(/\B(?=(\d{3})+(?!\d))/g, ',');
        var cents = rounded % 100;
        if (cents === 0) {
            return sign + whole;
        }
        return sign + whole + '.' + (cents < 10 ? '0' + cents : String(cents));
    });
}
"""#

private func intervalLabel(for interval: CandleInterval) -> String {
    interval == .oneDay ? "1D" : interval.displayName
}

private let chartIntervals: [CandleInterval] = [
    .oneMinute,
    .fiveMinutes,
    .oneHour,
    .fourHours,
    .oneDay,
]
